### PR TITLE
10.0 partner multi relation parent 

### PR DESCRIPTION
**Pull request by gfcapalbo**

### CODE REVIEW

File: partner_multi_relation_parent/hooks.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import api, SUPERUSER_ID
+
+def post_init_hook(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    partner_model = env['res.partner']
+    # get all fields with a parent
+    partners = partner_model.search([]).filtered('parent_id')
```
Comment (hbrunn) [inline partner_multi_relation_parent/hooks.py]: this is horribly inefficient. Search for your condition.

File: partner_multi_relation_parent/hooks.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    for partner in partners:
+        partner.update_relations(partner.parent_id.id)
```
Comment (hbrunn) [inline partner_multi_relation_parent/hooks.py]: why not call the function on the results directly?

File: partner_multi_relation_parent/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        # unlink previous relation
+        if old_parent_id:
+            previous = par_rel_mod.search(
+                [
+                    ('left_partner_id', '=', self.id),
+                    ('type_id', '=', type_relation),
+                    ('right_partner_id', '=', old_parent_id)
+                ]
+            )
+            previous.unlink()
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: it seems somewhat wasteful to me to unlink when you could just write the field

File: partner_multi_relation_parent/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        if not parent_id:
+            parent_id = self.parent_id.id
+        par_rel_mod = self.env['res.partner.relation']
+        type_relation = self.env.ref(
+            'partner_multi_relation_parent.parent_relation_type'
+        ).id
+        # unlink previous relation
+        if old_parent_id:
+            previous = par_rel_mod.search(
+                [
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: in such cases, collapse the opening and closing brackets:

```
.search([
    prop1,
    prop2,
])

File: partner_multi_relation_parent/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                    ('left_partner_id', '=', self.id),
+                    ('type_id', '=', type_relation),
+                    ('right_partner_id', '=', old_parent_id)
+                ]
+            )
+            previous.unlink()
+        # create new relations
+        par_rel_mod.create(
+            {'left_partner_id' : self.id,
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: same here

File: partner_multi_relation_parent/models/res_partner.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+             'type_id': type_relation,
+             'right_partner_id': parent_id,
+            }
+        )
+
+    @api.model
+    def create(self, vals):
+        res = super(ResPartner, self).create(vals=vals)
+        if "parent_id" in vals.keys():
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: no `.keys()`, `key in dict` is what we do in python

File: partner_multi_relation_parent/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            res.update_relations(None, vals['parent_id'])
+        return res
+
+    @api.multi
+    def write(self, vals , relation_create=True):
+        if relation_create:
+            for this in self:
+                if "parent_id" in vals.keys() and parent_id != False:
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: again. do you use a lint checker locally?

Comment (hbrunn) [reply partner_multi_relation_parent/models/res_partner.py]: and we don't compare with a boolean in python

File: partner_multi_relation_parent/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    @api.model
+    def create(self, vals):
+        res = super(ResPartner, self).create(vals=vals)
+        if "parent_id" in vals.keys():
+            res.update_relations(None, vals['parent_id'])
+        return res
+
+    @api.multi
+    def write(self, vals , relation_create=True):
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner.py]: don't change `write`'s signature, that's a path to misery. Use the context.

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+class ResPartnerRelation(models.Model):
+    _inherit = 'res.partner.relation'
+
+    @api.multi
+    def write(self, vals):
+        part_mod = self.env['res.partner']
+        for this in self:
+            type_relation = self.env.ref(
+                'partner_multi_relation_parent.parent_relation_type'
+            ).id
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: do lookups out of the loop

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            if this.type_id == type_relation:
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: invert the condition and continue

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                if ('type_id' in vals_keys()
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: this is going to raise

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+                        and vals['type_id'] != type_relation):
+                    part_mod.browse(this.left_partner_id).write(
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: why would you browse a perfectly good browse record again

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+                        {'parent_id' : False}
+                    )
+                    continue
+                if ('right_partner_id' in vals.keys() and 'left_partner_id' not
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: I'm not going to mention every `.keys()` call, fix them yourself

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                        in  vals.keys()):
+                    new_parent = vals.get('right_partner_id')
+                    contact_id = part_mod.browse(this.left_partner_id)
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: same about unnecessary `browse`s

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+            type_relation = self.env.ref(
+                'partner_multi_relation_parent.parent_relation_type'
+            ).id
+            if this.type_id == type_relation:
+                if ('type_id' in vals_keys()
+                        and vals['type_id'] != type_relation):
+                    part_mod.browse(this.left_partner_id).write(
+                        {'parent_id' : False}
+                    )
+                    continue
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: that's a case for `elif`

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+# © 2017 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from openerp import api, models
+
+
+class ResPartnerRelation(models.Model):
+    _inherit = 'res.partner.relation'
+
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py:10]: code for not allowing more than one parent relation, for unlinking a relation, for creating a relation is missing

File: partner_multi_relation_parent/models/res_partner_relation.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+                    old_contact_id = part_mod.browse(this.left_partner_id)
+                    old_contact_id.write(
+                        {'parent_id': False}, relation_create=False)
+                    contact_id = part_mod.browse(vals['left_partner_id'])
+                    contact_id.write(
+                        {'parent_id': vals['right_partner_id']},
+                        relation_create=False
+                    )
+        res = super(ResPartnerRelation, self).write(vals=vals)
+        return res
```
Comment (hbrunn) [inline partner_multi_relation_parent/models/res_partner_relation.py]: just return the function call directly in such a case

File: partner_multi_relation_parent/static/src/js/partner_multi_relation_parent.js [hunk 18]
```diff
@@ -0,0 +1,7 @@
+//-*- coding: utf-8 -*-
+//© 2017 Therp BV <http://therp.nl>
+//License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+openerp.partner_multi_relation_parent = function(instance)
+{
+};
```
Comment (hbrunn) [inline partner_multi_relation_parent/static/src/js/partner_multi_relation_parent.js]: clean up empty files, also the css above

File: partner_multi_relation_parent/tests/test_partner_multi_relation_parent.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+            ('right_partner_id', '=', parent_id)
+        ])
+        return len(hits)
+
+    def count_relations_of_parent(self, parent_id):
+        hits = self.par_rel_mod.search([
+            ('type_id' , '=', self.type_relation),
+            ('right_partner_id', '=', parent_id)
+        ])
+        return len(hits)
```
Comment (hbrunn) [inline partner_multi_relation_parent/tests/test_partner_multi_relation_parent.py]: `count=True` above is your friend

File: partner_multi_relation_parent/tests/test_partner_multi_relation_parent.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    def count_partner_relation_left_right(self, partner_id, parent_id=None):
+        # returns number relations , should be always 1
+        # todo extend this with relations_all  calculations (left, right)
+        if not parent_id:
+            return 0
+        hits =  self.par_rel_mod.search([
+            ('left_partner_id', '=', partner_id),
+            ('type_id' , '=', self.type_relation),
+            ('right_partner_id', '=', parent_id)
+        ])
```
Comment (hbrunn) [inline partner_multi_relation_parent/tests/test_partner_multi_relation_parent.py]: use `count=True`

File: partner_multi_relation_parent/views/templates.xml [hunk 21]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hbrunn) [inline partner_multi_relation_parent/views/templates.xml]: clean up unused files

**Review by hbrunn — CHANGES_REQUESTED** [review]
I also think this module should disable https://github.com/OCA/OCB/blob/10.0/odoo/addons/base/res/res_partner_view.xml#L343 to avoid confusion.

And don't forget to squash your commits before @daramousk merges

File: partner_multi_relation_parent/models/res_partner.py (hunk 7 above)
Comment (gfcapalbo) [reply partner_multi_relation_parent/models/res_partner.py]: #linter yes i have , and  it did not notify this poor form.

**Pull request closed without merging**